src/ts/ConvivaAnalytics.ts

```diff
@@ -657,79 +698,127 @@ export class ConvivaAnalytics {
-      // value and use it for tracking when initializing the session.
-      // TODO: remove this workaround when the player event order is fixed
-      this.lastSeenBitrate = bitrateKbps;
-      return;
-    }
-
-    this.lastSeenBitrate = null;
+    this.debugLog('[ ConvivaAnalytics ] report bitrate', {
+      event,
+      bitrateKbps,
+    });
     this.convivaVideoAnalytics.reportPlaybackMetric(Conviva.Constants.Playback.BITRATE, bitrateKbps);
   };
 
   private onCustomEvent = (event: PlayerEventBase) => {
+    this.debugLog('[ Player Event ] custom playback related event', event);
+
     if (!this.isSessionActive()) {
-      this.debugLog('skip custom event, no session existing', event);
+      this.debugLog('[ ConvivaAnalytics ] skip custom event, no session existing', event);
       return;
     }
 
     const eventAttributes = ObjectUtils.flatten(event);
     this.sendCustomPlaybackEvent(event.type, eventAttributes);
   };
 
-  private trackAdBreakStarted = (event: AdBreakEvent) => {
-    this.debugLog('[ ConvivaAnalytics ] adbreak started', event);
-    this.isAd = true;
+  // Fires before `onAdStarted`
+  private onAdBreakStarted = (event: AdBreakEvent) => {
+    this.debugLog('[ Player Event ] adbreak started', event);
 
-    const adPosition = AdBreakHelper.mapAdPosition(event.adBreak, this.player);
-
-    if (!this.isSessionActive()) {
-      // Don't report without a valid session (e.g., in case of a pre-roll, or post-roll ad)
-      return;
-    }
+    this.isAdBreak = true;
+    this.lastAdBreakEvent = event;
 
+    this.debugLog('[ ConvivaAnalytics ] report ad break started', event);
     this.convivaVideoAnalytics.reportAdBreakStarted(
       Conviva.Constants.AdType.CLIENT_SIDE,
       Conviva.Constants.AdPlayer.SEPARATE,
     );
   };
 
-  private onAdBreakFinished = (event: AdBreakEvent | ErrorEvent) => {
-    this.debugLog('[ ConvivaAnalytics ] adbreak finished', event);
-    this.isAd = false;
+  private onAdStarted = (event: AdEvent) => {
+    this.debugLog('[ Player Event ] ad started', event);
 
-    if (!this.isSessionActive()) {
-      // Don't report without a valid session (e.g., in case of a pre-roll, or post-roll ad)
-      return;
+    const adInfo = AdHelper.extractConvivaAdInfo(this.player, this.lastAdBreakEvent, event);
+    // It's already in kbps
+    const bitrateKbps = event.ad.data?.bitrate;
+
+    this.debugLog('[ ConvivaAdAnalytics ] report ad started', {
+      event,
+      adInfo,
+    });
+    this.convivaAdAnalytics.reportAdStarted(adInfo);
+
+    this.debugLog('[ ConvivaAdAnalytics ] report playing ad playback state');
+    this.convivaAdAnalytics.reportAdMetric(Conviva.Constants.Playback.PLAYER_STATE, Conviva.Constants.PlayerState.PLAYING);
+
+    if (bitrateKbps) {
+      this.debugLog('[ ConvivaAdAnalytics ] report ad bitrate', bitrateKbps);
+      this.convivaAdAnalytics.reportAdMetric(Conviva.Constants.Playback.BITRATE, bitrateKbps);
     }
+  }
+
+  // Fires before `onAdBreakFinished`
```

Comment:
I think this comment can be removed. This can be assumed general knowledge.

Reply:
Removed in https://github.com/bitmovin/bitmovin-player-web-analytics-conviva/pull/104/commits/763529964e102b9048dfb0aba21f7c1f56163bc1.